File: algorithms/380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.py
import random


class RandomizedSet:

    def __init__(self):
        self.nums = []
        self.indices = {}

    def insert(self, val: int) -> bool:
        if val in self.indices:
            return False
        self.indices[val] = len(self.nums)
        self.nums.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.indices:
            return False
        idx = self.indices[val]
        last = len(self.nums) - 1
        self.nums[idx] = self.nums[last]
        self.indices[self.nums[idx]] = idx
        self.nums = self.nums[:last]
        del self.indices[val]
        return True

    def getRandom(self) -> int:
        return random.choice(self.nums)

File: algorithms/380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.go
package problem380

import "math/rand"

// RandomizedSet 变长数组 + 哈希表
type RandomizedSet struct {
	nums    []int
	indices map[int]int
}

func Constructor() RandomizedSet {
	return RandomizedSet{
		[]int{}, map[int]int{},
	}
}

func (this *RandomizedSet) Insert(val int) bool {
	if _, ok := this.indices[val]; ok {
		return false
	}

	this.indices[val] = len(this.nums)
	this.nums = append(this.nums, val)
	return true
}

func (this *RandomizedSet) Remove(val int) bool {
	idx, ok := this.indices[val]
	if !ok {
		return false
	}
	last := len(this.nums) - 1
	this.nums[idx] = this.nums[last]
	this.indices[this.nums[idx]] = idx
	this.nums = this.nums[:last]
	delete(this.indices, val)
	return true
}

func (this *RandomizedSet) GetRandom() int {
	return this.nums[rand.Intn(len(this.nums))]
}
